diff --git a/dwl.c b/dwl.c
index b8ef17c..f493708 100644
--- a/dwl.c
+++ b/dwl.c
@@ -2279,7 +2279,7 @@ setup(void)
 	 * images are available at all scale factors on the screen (necessary for
 	 * HiDPI support). Scaled cursors will be loaded with each output. */
 	cursor_mgr = wlr_xcursor_manager_create(NULL, 24);
-	setenv("XCURSOR_SIZE", "24", 1);
+	setenv("XCURSOR_SIZE", "@cursorsize@", 1);
 
 	/*
 	 * wlr_cursor *only* displays an image on screen. It does not move around
