diff --git a/config.def.h b/config.def.h
index e4d1fc8..53f2fad 100644
--- a/config.def.h
+++ b/config.def.h
@@ -53,7 +53,7 @@ static const int repeat_delay = 600;
 static const int tap_to_click = 1;
 static const int tap_and_drag = 1;
 static const int drag_lock = 1;
-static const int natural_scrolling = 0;
+static const int natural_scrolling = @naturalscroll@;
 static const int disable_while_typing = 1;
 static const int left_handed = 0;
 static const int middle_button_emulation = 0;
